### PR TITLE
oop-special_class_method-medium-q5

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/special_class_method/medium/q5/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# oop-special_class_method-medium-q5
+
+## Question Text
+
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/QuestionText.md:4]: add something here or in MainTest describing what kind of question this is

File: src/oop/special_class_method/medium/q5/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+It's April fool's day and you want to change how the calculator works, so you can trick your Math teacher. Make a Java
+class called `Calculator` that has two attributes called **sum** and **isCorrect** whose values are initialized in the
+constructor. In particular, **sum** is set to the input parameter times 7. The class also has an accessor and mutator
+method for each attribute.
+
+### Calculator:
+
+Your `Calculator` class must have the following items
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/QuestionText.md:12]: change this to reflect that this stuff will be provided

File: src/oop/special_class_method/medium/q5/QuestionText.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+- Attributes:
+    - a `double` called **sum** that is not accessible from another class
+    - a `boolean` called **isCorrect** that is not accessible from another class
+
+- Constructor:
+    - a `Calculator` constructor that is visible to other classes and accepts a `double` and a `boolean` and uses it to
+      initialize the **sum** and **isCorrect** attributes with **sum** being 7 times of the input value.
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/QuestionText.md:20]:
```suggestion
      initialize the **sum** and **isCorrect** attributes with **sum** being 7 multiplied by the input value.
```

File: src/oop/special_class_method/medium/q5/QuestionText.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+You need to add the following items:
+
+- Methods:
+    - a `setSum` method which accepts a `double` and uses it to set the value of the **sum** attribute
+    - a `getSum` method which returns the value stored in the **sum** attribute
+    - a `setIsCorrect` method which accepts a `boolean` and uses it to set the value of the **isCorrect** attribute
+    - a `getIsCorrect` method which returns the value stored in the **isCorrect** attribute
+    - a `toString` method which returns the string `"The sum is " + sum + " and it is " + isCorrect`
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/QuestionText.md:29]: use x y formatting

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        Object[][] arguments = {
+                {value1, double.class},
+                {value2, boolean.class}
+        };
+        String incorrectToStringMessage = "Your toString method does not return the correct String.";
+        Object testInstance = testClass.createInstance(arguments);
+        String expectedOutput = "The sum is " + value1 * 7 + " and it is " + value2;
+        Object toStringOutput = testClass.callMethod("toString", testInstance);
+        _assertEquals(expectedOutput, toStringOutput, incorrectToStringMessage);
+    }
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/MainTest.java:173]: there should be a methods work together test here

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        Object[][] arguments = {
+                {sum, double.class},
+                {correct, boolean.class}
+        };
+        Object testInstance = testClass.createInstance(arguments);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName1, testInstance);
+        assertTrue(testClass.hasModifier(getAttributeMethodName1, null, "public"),
+                "Your " + getAttributeMethodName1 + " method does not have the correct visibility modifier.");
+        _assertEquals(sum * 7, getMethodOutput,
+                "Your " + getAttributeMethodName1 + " method does not return the value of the size attribute.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/MainTest.java:78]:
```suggestion
                "Your " + getAttributeMethodName1 + " method does not return the value of the " + attributeName1 + " attribute.");
```

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        Object[][] arguments = {
+                {value, double.class},
+                {on, boolean.class}
+        };
+        Object testInstance = testClass.createInstance(arguments);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName2, testInstance);
+        assertTrue(testClass.hasModifier(getAttributeMethodName2, null, "public"),
+                "Your " + getAttributeMethodName2 + " method does not have the correct visibility modifier.");
+        _assertEquals(on, getMethodOutput,
+                "Your " + getAttributeMethodName2 + " method does not return the value of the size attribute.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/MainTest.java:127]:
```suggestion
                "Your " + getAttributeMethodName2 + " method does not return the value of the " + attributeName2 + " attribute.");
```

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+                {c, boolean.class}
+        };
+        Object testInstance = testClass.createInstance(instantiationArguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, double.class}
+        };
+        Class<?>[] methodModifierClasses = {
+                double.class
+        };
+        Object setMethodOutput = testClass.callMethod(setAttributeMethodName1, setMethodArguments, testInstance);
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/MainTest.java:104]: We should probably have two separate tests for the method name and the arguments for better error messages

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+                {initialValue, boolean.class}
+        };
+        Object testInstance = testClass.createInstance(instantiationArguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, boolean.class}
+        };
+        Class<?>[] methodModifierClasses = {
+                boolean.class
+        };
+        Object setMethodOutput = testClass.callMethod(setAttributeMethodName2, setMethodArguments, testInstance);
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/MainTest.java:151]: this should also probably be split into two tests

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/medium/q5/QuestionText.md [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+      initialize the **sum** and **isCorrect** attributes with **sum** being 7 multiplied by the input value.
+
+You need to add the following items:
+
+- Methods:
+    - a `setSum` method which accepts a `double` and uses it to set the value of the **sum** attribute
+    - a `getSum` method which returns the value stored in the **sum** attribute
+    - a `setIsCorrect` method which accepts a `boolean` and uses it to set the value of the **isCorrect** attribute
+    - a `getIsCorrect` method which returns the value stored in the **isCorrect** attribute
+    - a `toString` method which returns the string `"The sum is " + x + " and it is " + y` where x is the value
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/QuestionText.md:29]:
```suggestion
    - a `toString` method which returns the string `"The sum is  x and it is y"` where x is the value
```

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+    }
+    @ParameterizedTest
+    @MethodSource("booleanDoublesInputProvider")
+    public void methodsWorkingTogether(double value1, boolean value2) throws Throwable {
+        Object[][] initialArguments = {
+                {1.23, double.class},
+                {true, boolean.class}
+        };
+        Object testInstance = testClass.createInstance(initialArguments);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName1, testInstance);
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/MainTest.java:187]: no need for 187-192, getting without first setting has already been tested previously

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        getMethodOutput = testClass.callMethod(getAttributeMethodName2, testInstance);
+        _assertEquals(true, getMethodOutput,
+                "Your " + getAttributeMethodName2 + " method does not return the value of the " + attributeName2 + " attribute.");
+        Object[][] setMethodArguments = {
+                {value1, double.class}
+        };
+        Object setMethodOutput = testClass.callMethod(setAttributeMethodName1, setMethodArguments, testInstance);
+        _assertEquals(value1, testClass.getFieldValue(testInstance, attributeName1),
+                "Your " + setAttributeMethodName1 + " method does not correctly update the value of " + attributeName1 + " after using the " + getAttributeMethodName1 + " method.");
+        assertNull(setMethodOutput, String.join(" ", "Your", setAttributeMethodName1, "method should not return any output"));
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q5/MainTest.java:199]: no need for the assertNull either since that will have been caught already

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by SamamthaH — DISMISSED** [review]

File: src/oop/special_class_method/medium/q5/QuestionText.md [hunk 13]
```diff
@@ -0,0 +1,31 @@
+# oop-special_class_method-medium-q5
+
+## Question Text
+
+It's April fool's day and you want to change how the calculator works, so you can trick your Math teacher. Make a Java
```
Comment (Hedgemon4) [inline src/oop/special_class_method/medium/q5/QuestionText.md:5]: Please remove this extra information

File: src/oop/special_class_method/medium/q5/QuestionText.md [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    - a `boolean` called **isCorrect** that is not accessible from another class
+
+- Constructor:
+    - a `Calculator` constructor that is visible to other classes and accepts a `double` and a `boolean` and uses it to
+      initialize the **sum** and **isCorrect** attributes with **sum** being 7 multiplied by the input value.
+
+You need to add the following items:
+
+- Methods:
+    - a `setSum` method which accepts a `double` and uses it to set the value of the **sum** attribute
```
Comment (Hedgemon4) [inline src/oop/special_class_method/medium/q5/QuestionText.md:25]: And specify that the setters should have a void return

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+
+    @ParameterizedTest
+    @MethodSource("booleanDoublesInputProvider")
+    public void correctGetSumMethod(double sum, boolean correct) throws Throwable {
+        Object[][] arguments = {
+                {sum, double.class},
+                {correct, boolean.class}
+        };
+        Object testInstance = testClass.createInstance(arguments);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName1, testInstance);
```
Comment (Hedgemon4) [inline src/oop/special_class_method/medium/q5/MainTest.java:75]: Please check that the method exists before calling it

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+        Object testInstance = testClass.createInstance(instantiationArguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, double.class}
+        };
+        Class<?>[] methodModifierClasses = {
+                double.class
+        };
+        assertTrue(testClass.hasMethod(setAttributeMethodName1, methodModifierClasses),
+                "Your " + setAttributeMethodName1 + " method does not have the correct arguments.");
+        assertTrue(testClass.hasModifier(setAttributeMethodName1, methodModifierClasses, "public"),
```
Comment (Hedgemon4) [inline src/oop/special_class_method/medium/q5/MainTest.java:107]: Check the return type using the hasReturnType assert before calling the method.

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+                "Your " + setAttributeMethodName1 + " method does not have the correct visibility modifier.");
+        Object setMethodOutput = testClass.callMethod(setAttributeMethodName1, setMethodArguments, testInstance);
+        String incorrectSetterMessage =
+                "Your " + setAttributeMethodName1 + " method does not correctly update the value of " + attributeName1 + ".";
+        _assertEquals(updatedValue, testClass.getFieldValue(testInstance, attributeName1),
+                incorrectSetterMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setAttributeMethodName1, "method should not return any output"));
```
Comment (Hedgemon4) [inline src/oop/special_class_method/medium/q5/MainTest.java:114]: Don't this approach

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+    @ParameterizedTest
+    @MethodSource("booleanDoublesInputProvider")
+    public void correctGetIsCorrectMethod(double value, boolean on) throws Throwable {
+        Object[][] arguments = {
+                {value, double.class},
+                {on, boolean.class}
+        };
+        Object testInstance = testClass.createInstance(arguments);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName2, testInstance);
+        assertTrue(testClass.hasModifier(getAttributeMethodName2, null, "public"),
```
Comment (Hedgemon4) [inline src/oop/special_class_method/medium/q5/MainTest.java:127]: Check method exists before calling it

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+                {value, double.class},
+                {initialValue, boolean.class}
+        };
+        Object testInstance = testClass.createInstance(instantiationArguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, boolean.class}
+        };
+        Class<?>[] methodModifierClasses = {
+                boolean.class
+        };
```
Comment (Hedgemon4) [inline src/oop/special_class_method/medium/q5/MainTest.java:153]: Check return type before calling it

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+        assertNull(setMethodOutput, String.join(" ", "Your", setAttributeMethodName2, "method should not return any output"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("booleanDoublesInputProvider")
+    public void correctToStringMethod(double value1, boolean value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, double.class},
+                {value2, boolean.class}
+        };
```
Comment (Hedgemon4) [inline src/oop/special_class_method/medium/q5/MainTest.java:172]: Check existence of method

File: src/oop/special_class_method/medium/q5/MainTest.java [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+        String incorrectToStringMessage = "Your toString method does not return the correct String.";
+        Object testInstance = testClass.createInstance(arguments);
+        String expectedOutput = "The sum is " + value1 * 7 + " and it is " + value2;
+        Object toStringOutput = testClass.callMethod("toString", testInstance);
+        _assertEquals(expectedOutput, toStringOutput, incorrectToStringMessage);
+    }
+    @ParameterizedTest
+    @MethodSource("booleanDoublesInputProvider")
+    public void methodsWorkingTogether(double value1, boolean value2) throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/medium/q5/MainTest.java:181]: For this test, I think it makes more sense to call the setter and then the getter

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**